src/azul/service/async_manifest_service.py

```diff
@@ -63,40 +78,80 @@ class AsyncManifestService(AbstractService):
-                                                  execution_input=input)
-        return Token(execution_id=execution_id,
-                     request_index=0,
-                     wait_time=self._get_next_wait_time(0))
 
-    def inspect_generation(self, token) -> Union[Token, JSON]:
-        try:
-            execution = self.step_function_helper.describe_execution(state_machine_name=self.state_machine_name,
-                                                                     execution_name=token.execution_id)
-        except ClientError as e:
-            if e.response['Error']['Code'] == 'ExecutionDoesNotExist':
-                raise InvalidTokenError from e
+        execution_id = input['object_key'].rpartition('/')[2]
+        execution_id = execution_id.partition('.')[0]
+        token = Token(execution_id=f'{execution_id}_0',
+                      request_index=0,
+                      wait_time=self._get_next_wait_time(0))
+        for retry_index in range(self.step_function_execution_retries):
+            try:
+                self.step_function_helper.start_execution(self.state_machine_name,
+                                                          token.execution_id,
+                                                          execution_input=input)
+            except ClientError as e:
+                if e.response['Error']['Code'] == 'ExecutionAlreadyExists':
+                    output, status = self._inspect_generation(token)
+                    if status not in Status:
+                        raise StateMachineError(status, output)
+                    if status is Status.running:
+                        token = token.advance(wait_time=self._get_next_wait_time(token.request_index))
+                        break
+                    else:
+                        # If a manifest expires before the name of the Step
+                        # Function execution that created it can be reused
+                        # (90 days), StartExecution will fail with
+                        # ExecutionAlreadyExists. In that case we need to retry
+                        # even a successful execution, so that it can recreate
+                        # the manifest.
+                        retry_index += 1
+                        token = Token(execution_id=f'{execution_id}_{retry_index}',
+                                      request_index=0,
+                                      wait_time=self._get_next_wait_time(0))
+                else:
+                    raise
             else:
-                raise
-        output = execution.get('output', None)
-        status = execution['status']
-        if status == 'SUCCEEDED':
+                break
+        if token.execution_id.endswith(str(self.step_function_execution_retries)):
+            raise StateMachineError('Maximum manifest generation retries reached.',
+                                    token.execution_id)
+        else:
+            return token
```

Comment:
There must be a better way to structure the control flow. Like, using an `else:` branch on the loop or returning directly from the body of the loop.